### ISSUE TITLE
cachebust_match argument not documented; not allowed by add_static_view

### ISSUE THREAD

**Issue by ztane** [Pylons/pyramid#2675]
We are generating one path segment cachebreaker for _all_ of our static files (our static files do not really change that often, only in release cycles, thus it is OK to reload all of them whenever a new release occurs); this allows us to use relative urls without manifests and allows things like require.js work painlessly as well; but for this we utilized a modified `PathSegmentCacheBuster`.

Now that the new cachebusting scheme has come, it is not possible to do this any longer, since `add_static_view` doesn't like seeing cache tokens any more.

`static_view` seems to support `cachebust_match` which could be used to remove the busting token from the subpath, but `add_static_view` thinks an argument by that name is a custom predicate instead.

**Comment by ztane** [comment]
And if it wasn't clear, what we've been doing is serving all the static files at

     /static/<application-git-revision-id>/

so that the relative urls in CSS and more importantly *JavaScript* as we're using on-demand loading for various frameworks, including loading locales for libraries and such, can continue to work with relative URLs, and we have to rewrite *absolutely* *nothing* to get cache busting work correctly everywhere, and every file, including that 16x16 pixel logo added that included css will also be cachebusted, ensuring that nothing needs to be flushed in CDN.

**Comment by bertjwregeer** [comment]
Unfortunately we had to remove the ability for the static views to remove cache busting tokens because it is impossible to do so sanely.

See my comment here: https://github.com/Pylons/pyramid/pull/2186#issuecomment-166031512

Also see this comment: https://github.com/Pylons/pyramid/pull/2171#issuecomment-163728746

I am not sure why `cachebust_match` still exists. Maybe @mmerickel can shed some light on that, but at this point in time there is no support for going from a cache buster URL to a non-cache busted asset spec.

Your best bet is to write a tween that re-writes the URL on the way in.

**Comment by bertjwregeer** [comment]
Confirmed with @mmerickel on IRC that `cachebust_match` was left-over from when that feature was ripped out, and needs to be excised from the codebase.

**Comment by mmerickel** [comment]
That feature was intended to be removed, however it looks like the static_view support slipped through the cracks. That is not to say that your use-case is invalid, or shouldn't be discussed.

The main reason this is not solved generally in Pyramid (right now) is that we decided not to make any assumptions about the URL that would be generated by your cachebuster. This means it may not be mounted in `/static` (the url that is being matched by `add_static_view`). The cache buster handles ONLY generating busted urls and we left the matching up to the user as we have no idea how the urls will be generated. The way to solve this (right now) is to add a tween which can match busted urls and rewrite them back to their original version. This is easy to do for a case such as yours. Here is an example that I have used in a real app to match urls of the format `/static/<asset>-<token>.<ext>`. https://gist.github.com/mmerickel/61023515229b6645e80d

I hope this will a) help you solve your problem and b) give you some fodder for discussing this further if we are to solve it in Pyramid. The comments linked above by @bertjwregeer are still relevant.

**Comment by bertjwregeer** [comment]
I've created a PR that removes `cachebust_match`: https://github.com/Pylons/pyramid/pull/2681

**Comment by mmerickel** [comment]
@ztane Your issue is easily solved by simply not using the cache busting framework. A global cache busting token is quite trivial:

```python
config.add_static_view('static/' + token, 'myapp:static', http_cache=10 * 365 * 24 * 3600)

# let's redirect other requests to tokenized versions
config.add_route('redirect-static', '/static/*subpath')
def static_redirect_view(request):
    next_url = request.static_url('myapp:static/' + '/'.join(request.subpath), _query=request.GET)
    return HTTPFound(next_url)
config.add_view(static_redirect_view, route_name='redirect-static')
```

**Comment by ztane** [comment]
Yeah that was what we were using in pyramid 1.5.

**Comment by mmerickel** [comment]
We merged #2681 which removes this undocumented argument. You'll need to come up with another way to serve the cache busted urls. Right now pyramid only helps generate them. Several possible solutions were mentioned already in this issue.
